pkg/providers/ibmcloud.go

```diff
@@ -0,0 +1,331 @@
+	"github.com/softlayer/softlayer-go/session"
+	"github.com/softlayer/softlayer-go/sl"
+
+	"github.com/openshift/ofcir/pkg/utils"
+)
+
+type ibmcloudProviderConfig struct {
+	APIKey string `json:"apikey"`
+	Sshkey string `json:"sshkey"`
+	Preset string `json:"preset"`
+}
+
+type ibmcloudProvider struct {
+	config ibmcloudProviderConfig
+	client *session.Session
+}
+
+func IbmcloudProviderFactory(providerInfo string, secretData map[string][]byte) (Provider, error) {
+	config := ibmcloudProviderConfig{}
+
+	if configJSON, ok := secretData["config"]; ok {
+		if err := json.Unmarshal(configJSON, &config); err != nil {
+			return nil, fmt.Errorf("error in provider config json: %w", err)
+		}
+	}
+
+	provider := &ibmcloudProvider{
+		config: config,
+	}
+
+	provider.client = session.New("apikey", provider.config.APIKey)
+	return provider, nil
+}
+
+func (p *ibmcloudProvider) getPackageItems(packageId int) (resp []datatypes.Product_Item) {
+	var mask = "id, itemCategory, keyName,prices[id, hourlyRecurringFee, recurringFee, categories]"
+	var service = services.GetProductPackageService(p.client)
+	receipt, err := service.Id(packageId).Mask(mask).GetItems()
+	if err != nil {
+		return
+	}
+
+	return receipt
+}
+
+func (p *ibmcloudProvider) getStandardPrice(item datatypes.Product_Item) (resp datatypes.Product_Item_Price) {
+	for _, itemPrice := range item.Prices {
+		if itemPrice.LocationGroupId == nil {
+			return itemPrice
+		}
+	}
+	return datatypes.Product_Item_Price{}
+
+}
+
+func (p *ibmcloudProvider) getItemPriceList(packageId int, itemKeyNames []string) (resp []datatypes.Product_Item_Price) {
+
+	items := p.getPackageItems(packageId)
+	var prices []datatypes.Product_Item_Price
+
+	for _, itemKeyName := range itemKeyNames {
+		for _, item := range items {
+			if (*item.KeyName) == itemKeyName {
+				itemPrice := p.getStandardPrice(item)
+				prices = append(prices, itemPrice)
+				break
+			}
+		}
+	}
+
+	return prices
+}
+
+func (p *ibmcloudProvider) serverCreate(pool, presetname string) (string, error) {
+
+	// TODO: Would love to find a way to list locations where the package is available but
+	// I can't for the life of me figure it out, need to revisit
+	// I've removed some of the more expensive locations from the list below
+	locations := []string{"DALLAS09", "DALLAS10", "DALLAS12", "DALLAS13", "FRANKFURT", "FRANKFURT04", "FRANKFURT05", "LONDON02", "LONDON04",
+		"LONDON05", "LONDON06", "MILAN", "MONTREAL", "PARIS", "MADRID02", "MADRID04", "MADRID05", "SANJOSE03", "SANJOSE04", "TORONTO",
+		"TORONTO04", "TORONTO05", "WASHINGTON06", "WASHINGTON07"}
+	rand.Shuffle(len(locations), func(i, j int) {
+		locations[i], locations[j] = locations[j], locations[i]
+	})
+
+	getLocation := func() (string, error) {
+		if len(locations) == 0 {
+			return "", errors.New("something went wrong")
+		}
+		return locations[0], nil
+	}
+	location, err := getLocation()
+
+	// Package 200 are the fast provisioning servers (40-60min provsion)
+	packageId := 200
```

Comment:
Also here a speaking const, ie `const IbmCloudFastProvisioningServer = 200` could help in the maintenance

Reply:
will do